File: src/main/java/hiiragi283/material/HiiragiItems.java
package hiiragi283.material;

import hiiragi283.material.api.item.HiiragiItem;
import hiiragi283.material.api.item.MaterialItem;
import hiiragi283.material.api.material.HiiragiMaterial;
import hiiragi283.material.api.registry.HiiragiEntry;
import hiiragi283.material.api.shape.HiiragiShapes;
import hiiragi283.material.api.shape.ShapeType;
import hiiragi283.material.config.RMConfig;
import hiiragi283.material.item.ItemBookRespawn;
import hiiragi283.material.util.CraftingBuilder;
import hiiragi283.material.util.HiiragiIngredient;
import hiiragi283.material.util.HiiragiUtil;
import net.minecraft.client.renderer.block.model.ModelResourceLocation;
import net.minecraft.client.renderer.color.ItemColors;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraftforge.client.model.ModelLoader;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import net.minecraftforge.registries.IForgeRegistry;

import java.util.ArrayList;
import java.util.List;

public class HiiragiItems implements HiiragiEntry.ITEM {

    public static HiiragiItems INSTANCE = new HiiragiItems();

    private HiiragiItems() {
    }

    private static final List<ITEM> ENTRIES = new ArrayList<>();

    //    Items    //

    public static final HiiragiItem BOOK_RESPAWN = new ItemBookRespawn();

    public static final MaterialItem MATERIAL_BLOCK = MaterialItem.create(
            HiiragiShapes.BLOCK,
            (item, material) -> {
                if (material.isMetal()) {
                    new CraftingBuilder(item.asItemStack(material))
                            .setPattern("AAA", "AAA", "AAA")
                            .setIngredient('A', HiiragiShapes.INGOT.getOreDict(material))
                            .build();
                } else if (material.isGem()) {
                    new CraftingBuilder(item.asItemStack(material))
                            .setPattern("AAA", "AAA", "AAA")
                            .setIngredient('A', HiiragiShapes.GEM.getOreDict(material))
                            .build();
                }
            }
    );

    public static final MaterialItem MATERIAL_BOTTLE = MaterialItem.create(HiiragiShapes.BOTTLE);

    public static final MaterialItem MATERIAL_DUST = MaterialItem.create(
            HiiragiShapes.DUST,
            (item, material) -> new CraftingBuilder(item.asItemStack(material))
                    .setPattern("AAA", "AAA", "AAA")
                    .setIngredient('A', HiiragiShapes.DUST_TINY.getOreDict(material))
                    .build()
    );

    public static final MaterialItem MATERIAL_DUST_TINY = MaterialItem.create(
            HiiragiShapes.DUST_TINY,
            (item, material) -> new CraftingBuilder(item.asItemStack(material, 9))
                    .addIngredient(new HiiragiIngredient(HiiragiShapes.DUST.getOreDict(material)))
                    .build()
    );

    public static final MaterialItem MATERIAL_GEAR = MaterialItem.create(HiiragiShapes.GEAR);

    public static final MaterialItem MATERIAL_GEM = MaterialItem.create(
            HiiragiShapes.GEM,
            item -> {
                ModelLoader.registerItemVariants(item.asItem(),
                        HiiragiUtil.appendPath(item.getLocation(), "_" + ShapeType.GEM_AMORPHOUS.name()),
                        HiiragiUtil.appendPath(item.getLocation(), "_" + ShapeType.GEM_COAL.name()),
                        HiiragiUtil.appendPath(item.getLocation(), "_" + ShapeType.GEM_CUBIC.name()),
                        HiiragiUtil.appendPath(item.getLocation(), "_" + ShapeType.GEM_DIAMOND.name()),
                        HiiragiUtil.appendPath(item.getLocation(), "_" + ShapeType.GEM_EMERALD.name()),
                        HiiragiUtil.appendPath(item.getLocation(), "_" + ShapeType.GEM_LAPIS.name()),
                        HiiragiUtil.appendPath(item.getLocation(), "_" + ShapeType.GEM_QUARTZ.name()),
                        HiiragiUtil.appendPath(item.getLocation(), "_" + ShapeType.GEM_RUBY.name())
                );
                ModelLoader.setCustomMeshDefinition(item.asItem(), stack -> {
                    var shapeType = HiiragiMaterial.REGISTRY_INDEX.getValue(stack.getMetadata()).shapeType();
                    return new ModelResourceLocation(HiiragiUtil.appendPath(item.getLocation(), "_" + shapeType.name()), "inventory");
                });
            },
            (item, material) -> new CraftingBuilder(item.asItemStack(material, 9))
                    .addIngredient(new HiiragiIngredient(HiiragiShapes.BLOCK.getOreDict(material)))
                    .build()
    );

    public static final MaterialItem MATERIAL_INGOT = MaterialItem.create(
            HiiragiShapes.INGOT,
            (item, material) -> {
                new CraftingBuilder(item.asItemStack(material))
                        .setPattern("AAA", "AAA", "AAA")
                        .setIngredient('A', HiiragiShapes.NUGGET.getOreDict(material))
                        .build();
                ItemStack ingot9 = item.asItemStack(material, 9);
                new CraftingBuilder(HiiragiUtil.appendPath(HiiragiUtil.toLocation(ingot9, '_'), "_alt"), ingot9)
                        .addIngredient(new HiiragiIngredient(HiiragiShapes.BLOCK.getOreDict(material)))
                        .build();
            }
    );

    public static final MaterialItem MATERIAL_NUGGET = MaterialItem.create(
            HiiragiShapes.NUGGET,
            (item, material) -> new CraftingBuilder(item.asItemStack(material, 9))
                    .addIngredient(new HiiragiIngredient(HiiragiShapes.INGOT.getOreDict(material)))
                    .build()
    );

    public static final MaterialItem MATERIAL_PLATE = MaterialItem.create(HiiragiShapes.PLATE);

    public static final MaterialItem MATERIAL_STICK = MaterialItem.create(HiiragiShapes.STICK);

    public static final MaterialItem MATERIAL_WIRE = MaterialItem.create(HiiragiShapes.WIRE);

    //    Registry    //

    public static void init() {
        RagiMaterials.LOGGER.info("HiiragiItems has been initialized!");
        ENTRIES.addAll(HiiragiBlocks.getItemBlocks());
        ENTRIES.add(BOOK_RESPAWN);
        if (!RMConfig.EXPERIMENTAL.enableMetaTileBlock) ENTRIES.add(MATERIAL_BLOCK);
        ENTRIES.add(MATERIAL_BOTTLE);
        ENTRIES.add(MATERIAL_DUST);
        ENTRIES.add(MATERIAL_DUST_TINY);
        ENTRIES.add(MATERIAL_GEAR);
        ENTRIES.add(MATERIAL_GEM);
        ENTRIES.add(MATERIAL_INGOT);
        ENTRIES.add(MATERIAL_NUGGET);
        ENTRIES.add(MATERIAL_PLATE);
        ENTRIES.add(MATERIAL_STICK);
    }

    @Override
    public void register(IForgeRegistry<Item> registry) {
        ENTRIES.forEach(item -> registry.register(item.getObject()));
    }

    @Override
    public void registerOreDict() {
        ENTRIES.forEach(HiiragiEntry::registerOreDict);
    }

    @Override
    public void registerRecipe() {
        ENTRIES.forEach(HiiragiEntry::registerRecipe);
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void registerItemColor(ItemColors itemColors) {
        ENTRIES.forEach(item -> item.registerItemColor(itemColors));
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void registerModel() {
        ENTRIES.forEach(HiiragiEntry::registerModel);
    }

}

File: src/main/kotlin/hiiragi283/material/HiiragiItems.kt
package hiiragi283.material

import hiiragi283.material.api.item.MaterialItem
import hiiragi283.material.api.material.HiiragiMaterial
import hiiragi283.material.api.registry.HiiragiEntry
import hiiragi283.material.api.registry.HiiragiRegistries
import hiiragi283.material.api.shape.HiiragiShapeType
import hiiragi283.material.api.shape.HiiragiShapeTypes
import hiiragi283.material.api.shape.HiiragiShapes
import hiiragi283.material.config.HiiragiConfigs
import hiiragi283.material.item.*
import hiiragi283.material.util.*
import net.minecraft.client.renderer.block.model.ModelResourceLocation
import net.minecraft.item.ItemStack
import net.minecraft.item.crafting.Ingredient
import net.minecraftforge.client.model.ModelLoader
import net.minecraftforge.common.crafting.CraftingHelper

object HiiragiItems : HiiragiEntry.ITEM {

    @JvmField
    val BOOK_RESPAWN = HiiragiRegistries.ITEM.register(ItemBookRespawn)

    //    Material    //

    @JvmField
    val MATERIAL_BLOCK = HiiragiRegistries.ITEM.registerOptional(
        MaterialItem(
            HiiragiShapes.BLOCK,
            recipe = { item: MaterialItem, material: HiiragiMaterial ->
                if (HiiragiShapes.INGOT.isValid(material)) {
                    CraftingBuilder(item.itemStack(material))
                        .setPattern("AAA", "AAA", "AAA")
                        .setIngredient('A', HiiragiShapes.INGOT.getOreDict(material))
                        .build()
                } else if (HiiragiShapes.GEM.isValid(material)) {
                    CraftingBuilder(item.itemStack(material))
                        .setPattern("AAA", "AAA", "AAA")
                        .setIngredient('A', HiiragiShapes.GEM.getOreDict(material))
                        .build()
                }
            }
        )
    ) { !HiiragiConfigs.EXPERIMENTAL.enableMetaTileBlock }

    @JvmField
    val MATERIAL_BOTTLE = HiiragiRegistries.ITEM.register(MaterialItem(HiiragiShapes.BOTTLE))

    @JvmField
    val MATERIAL_CASING = HiiragiRegistries.ITEM.registerOptional(MaterialItemCasing)
    { !HiiragiConfigs.EXPERIMENTAL.enableMetaTileBlock }

    @JvmField
    val MATERIAL_DUST = HiiragiRegistries.ITEM.register(MaterialItem(HiiragiShapes.DUST))

    @JvmField
    val MATERIAL_GEAR = HiiragiRegistries.ITEM.register(MaterialItem(
        HiiragiShapes.GEAR,
        recipe = { item: MaterialItem, material: HiiragiMaterial ->
            if (!HiiragiShapes.INGOT.isValid(material)) return@MaterialItem
            CraftingBuilder(item.itemStack(material))
                .setPattern(" A ", "ABA", " A ")
                .setIngredient('A', HiiragiShapes.INGOT.getOreDict(material))
                .setIngredient('B', WRENCH, true)
                .build()
        }
    ))

    @JvmField
    val MATERIAL_GEM = HiiragiRegistries.ITEM.register(
        MaterialItem(
        HiiragiShapes.GEM,
            model = { item: MaterialItem ->

            ModelLoader.registerItemVariants(
                item,
                item.registryName!!.append("_" + HiiragiShapeTypes.GEM_AMORPHOUS.name),
                item.registryName!!.append("_" + HiiragiShapeTypes.GEM_COAL.name),
                item.registryName!!.append("_" + HiiragiShapeTypes.GEM_CUBIC.name),
                item.registryName!!.append("_" + HiiragiShapeTypes.GEM_DIAMOND.name),
                item.registryName!!.append("_" + HiiragiShapeTypes.GEM_EMERALD.name),
                item.registryName!!.append("_" + HiiragiShapeTypes.GEM_LAPIS.name),
                item.registryName!!.append("_" + HiiragiShapeTypes.GEM_QUARTZ.name),
                item.registryName!!.append("_" + HiiragiShapeTypes.GEM_RUBY.name)
            )

            ModelLoader.setCustomMeshDefinition(item) { stack: ItemStack ->
                HiiragiRegistries.MATERIAL_INDEX.getValue(stack.metadata)?.shapeType
                    ?.let { shapeType: HiiragiShapeType ->
                        ModelResourceLocation(
                            item.registryName!!.append("_" + shapeType.name), "inventory"
                        )
                    }
                    ?: ModelResourceLocation(item.registryName!!, "inventory")
            }

        },
            recipe = { entry: MaterialItem, material: HiiragiMaterial ->
            if (!HiiragiShapes.BLOCK.isValid(material)) return@MaterialItem
                CraftingBuilder(entry.itemStack(material, 9))
                .addIngredient(CraftingHelper.getIngredient(HiiragiShapes.BLOCK.getOreDict(material)))
                .build()
        }
    ))

    @JvmField
    val MATERIAL_INGOT = HiiragiRegistries.ITEM.register(
        MaterialItem(
        HiiragiShapes.INGOT,
            recipe = { item: MaterialItem, material: HiiragiMaterial ->
            //nugget -> ingot
            if (!HiiragiShapes.NUGGET.isValid(material)) return@MaterialItem
                CraftingBuilder(item.itemStack(material))
                .setPattern("AAA", "AAA", "AAA")
                .setIngredient('A', HiiragiShapes.NUGGET.getOreDict(material))
                .build()
            //block -> ingot
            if (!HiiragiShapes.BLOCK.isValid(material)) return@MaterialItem
                val ingot9 = item.itemStack(material, 9)
            CraftingBuilder(ingot9.toLocation("_").append("_alt"), ingot9)
                .addIngredient(CraftingHelper.getIngredient(HiiragiShapes.BLOCK.getOreDict(material)))
                .build()
        }
    ))

    @JvmField
    val MATERIAL_NUGGET = HiiragiRegistries.ITEM.register(
        MaterialItem(
        HiiragiShapes.NUGGET,
            recipe = { item: MaterialItem, material: HiiragiMaterial ->
            if (!HiiragiShapes.INGOT.isValid(material)) return@MaterialItem
                CraftingBuilder(item.itemStack(material, 9))
                .addIngredient(CraftingHelper.getIngredient(HiiragiShapes.INGOT.getOreDict(material)))
                .build()
        }
    ))

    @JvmField
    val MATERIAL_PLATE: MaterialItem = HiiragiRegistries.ITEM.register(MaterialItem(
        HiiragiShapes.PLATE,
        recipe = { item: MaterialItem, material: HiiragiMaterial ->
            if (!HiiragiShapes.INGOT.isValid(material)) return@MaterialItem
            CraftingBuilder(item.itemStack(material))
                .addIngredient(CraftingHelper.getIngredient(HiiragiShapes.INGOT.getOreDict(material)))
                .addIngredient(Ingredient.fromStacks(WRENCH.itemStackWild()))
                .build()
        }
    ))

    @JvmField
    val MATERIAL_STICK: MaterialItem = HiiragiRegistries.ITEM.register(MaterialItem(
        HiiragiShapes.STICK,
        recipe = { item: MaterialItem, material: HiiragiMaterial ->
            if (!HiiragiShapes.INGOT.isValid(material)) return@MaterialItem
            CraftingBuilder(item.itemStack(material))
                .setPattern("AB", "A ")
                .setIngredient('A', HiiragiShapes.INGOT.getOreDict(material))
                .setIngredient('B', WRENCH, true)
                .build()
        }
    ))

    //    Module Item   //

    @JvmField
    val MODULE_MOTOR = HiiragiRegistries.ITEM.register(ItemMotor)

    //    Common    //

    @JvmField
    val MINECART_TANK = HiiragiRegistries.ITEM.register(ItemMinecartTank)

    @JvmField
    val SHAPE_PATTERN = HiiragiRegistries.ITEM.register(ItemShapePattern)

    @JvmField
    val WRENCH = HiiragiRegistries.ITEM.register(ItemWrench)

}